bin/scale_stack.py

```diff
@@ -33,38 +48,96 @@ def scale_stack(args):
+
+    if args.mode == 'up':
+        print("Starting bastion")
+        if not args.asg_only:
+            ec2_client.start_instances(InstanceIds=[instance_ids[f"bastion.{args.bosslet_name}"]])
+            time.sleep(1)
+
+        print("Waiting for bastion to initialize")
+        wait_for_instance(ec2_client, f"bastion.{args.bosslet_name}")
+
+        print("Scaling up vault")
+        asg_client.update_auto_scaling_group(AutoScalingGroupName=bosslet_asg['vault'], MinSize=1, MaxSize=1, DesiredCapacity=1)
+        time.sleep(1)
 
-    # Adjust desired, max, and min capacity for matched ASGs
-    for asg in bosslet_asg:
-        if args.mode == 'up':
-            client.update_auto_scaling_group(AutoScalingGroupName=asg['AutoScalingGroupName'], MinSize=1, MaxSize=1, DesiredCapacity=1)
-        else: 
-            client.update_auto_scaling_group(AutoScalingGroupName=asg['AutoScalingGroupName'], MinSize=0, MaxSize=0, DesiredCapacity=0)
+        print("Waiting for vault to initialize")
+        wait_for_instance(ec2_client, f"vault.{args.bosslet_name}")
 
+        print("Scaling up endpoint") 
+        asg_client.update_auto_scaling_group(AutoScalingGroupName=bosslet_asg['endpoint'], MinSize=1, MaxSize=1, DesiredCapacity=1)
+        time.sleep(1)
+        
+        print("Scaling up activities")
+        asg_client.update_auto_scaling_group(AutoScalingGroupName=bosslet_asg['activities'], MinSize=1, MaxSize=1, DesiredCapacity=1)
+        time.sleep(1)
+
+        print("Starting cachemanager")
+        if not args.asg_only:
+            ec2_client.start_instances(InstanceIds=[instance_ids[f"cachemanager.{args.bosslet_name}"]])
+            time.sleep(1)          
+    else:
+        print("Stopping cachemanager")
+        if not args.asg_only:
+            ec2_client.stop_instances(InstanceIds=[instance_ids[f"cachemanager.{args.bosslet_name}"]])
+            time.sleep(1)
+
+        print("Scaling down activties")
+        asg_client.update_auto_scaling_group(AutoScalingGroupName=bosslet_asg['activities'], MinSize=0, MaxSize=0, DesiredCapacity=0)
+        time.sleep(1)
+
+        print("Scaling down endpoint") 
+        asg_client.update_auto_scaling_group(AutoScalingGroupName=bosslet_asg['endpoint'], MinSize=0, MaxSize=0, DesiredCapacity=0)
+        time.sleep(1)
+
+        print("Scaling down vault")
+        asg_client.update_auto_scaling_group(AutoScalingGroupName=bosslet_asg['vault'], MinSize=0, MaxSize=0, DesiredCapacity=0)
+        time.sleep(1)
+
+        print("Stopping bastion")
+        if not args.asg_only:
+            ec2_client.stop_instances(InstanceIds=[instance_ids[f"bastion.{args.bosslet_name}"]])
+            time.sleep(1)
+        
     
+    print('Done!')
+
+def get_instance_id(ec2_client, instance_name):
+    resp = ec2_client.describe_instances(Filters=[{"Name":"tag:Name", "Values": [instance_name]}])['Reservations']
+    return resp[0]['Instances'][0]['InstanceId']
+
+def wait_for_instance(ec2_client, instance_name):
```

Comment:
boto3 has a waiter for this: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Waiter.InstanceRunning